### ISSUE TITLE
[clang-vg-autotest] No space left on device

### ISSUE THREAD

**Issue by KKaras169169** [spdk/spdk#2115]
# CI Intermittent Failure

CI failed during **clang-vg-autotest**, inside **unittest_build** test with error: 
_fatal error: error in backend: IO failure on output stream: No space left on device_

## Link to the failed CI build

https://ci.spdk.io/results/autotest-per-patch/builds/56527/archive/clang-vg-autotest/build.log

## Execution failed at

00:06:23.326    CXX test/cpp_headers/nvmf_fc_spec.o
_**00:06:24.077  fatal error: error in backend: IO failure on output stream: No space left on device**_
00:06:24.077  PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
00:06:24.077  Stack dump:
00:06:24.077  0.	Program arguments: /usr/bin/clang++ -o nvmf_fc_spec.o -MMD -MP -MF nvmf_fc_spec.d.tmp -g -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wmissing-declarations -fno-strict-aliasing -I/home/vagrant/spdk_repo/spdk/include -march=native -Werror -Wformat -Wformat-security -D_GNU_SOURCE -fPIC -fstack-protector -fno-common -I/home/vagrant/spdk_repo/spdk/isa-l/.. -DDEBUG -O0 -fno-omit-frame-pointer -fprofile-arcs -ftest-coverage -fsanitize=address -fsanitize=undefined -DSPDK_GIT_COMMIT=634461094 -pthread -c nvmf_fc_spec.cpp 
00:06:24.077  1.	<eof> parser at end of file
00:06:24.077  2.	Per-module optimization passes
00:06:24.077  3.	Running pass 'GCOV Profiler' on module 'nvmf_fc_spec.cpp'.
00:06:24.080  /usr/bin/ld.bfd: final link failed: No space left on device
00:06:24.103  clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
00:06:24.106  make[3]: *** [/home/vagrant/spdk_repo/spdk/mk/spdk.app.mk:71: err_injection] Error 1
00:06:24.106  make[2]: *** [/home/vagrant/spdk_repo/spdk/mk/spdk.subdirs.mk:44: err_injection] Error 2
00:06:24.106  make[2]: *** Waiting for unfinished jobs....
00:06:24.110    CXX test/cpp_headers/nvmf.o
00:06:24.143  /usr/lib64/libLLVM-11.so(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamE+0x2a)[0x7f66fa20427a]
00:06:24.143  /usr/lib64/libLLVM-11.so(_ZN4llvm3sys17RunSignalHandlersEv+0x34)[0x7f66fa2024e4]
00:06:24.144  /usr/lib64/libLLVM-11.so(_ZN4llvm20CrashRecoveryContext10HandleExitEi+0x74)[0x7f66fa138cc4]
00:06:24.144  /usr/lib64/libLLVM-11.so(_ZN4llvm3sys7Process4ExitEi+0x17)[0x7f66fa1fd0b7]
00:06:24.144  /usr/bin/clang++(+0x16162)[0x555d416fa162]
00:06:24.144  /usr/lib64/libLLVM-11.so(_ZN4llvm18report_fatal_errorERKNS_5TwineEb+0x78)[0x7f66fa147428]
00:06:24.145  /usr/lib64/libLLVM-11.so(_ZN4llvm18report_fatal_errorENS_9StringRefEb+0x0)[0x7f66fa1475a0]
00:06:24.145  /usr/lib64/libLLVM-11.so(_ZN4llvm14raw_fd_ostreamD1Ev+0x109)[0x7f66fa1e4029]
00:06:24.145  /usr/lib64/libLLVM-11.so(+0x1534529)[0x7f66facad529]
00:06:24.145  /usr/lib64/libLLVM-11.so(+0x1536e88)[0x7f66facafe88]
00:06:24.146  /usr/lib64/libLLVM-11.so(+0x1536ef2)[0x7f66facafef2]
00:06:24.146  /usr/lib64/libLLVM-11.so(_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE+0x33d)[0x7f66fa31955d]
00:06:24.146  /usr/lib64/libclang-cpp.so.11(_ZN5clang17EmitBackendOutputERNS_17DiagnosticsEngineERKNS_19HeaderSearchOptionsERKNS_14CodeGenOptionsERKNS_13TargetOptionsERKNS_11LangOptionsERKN4llvm10DataLayoutEPNSE_6ModuleENS_13BackendActionESt10unique_ptrINSE_17raw_pwrite_streamESt14default_deleteISM_EE+0x1447)[0x7f66fff0e887]
00:06:24.146  /usr/lib64/libclang-cpp.so.11(+0x172d3d2)[0x7f670019e3d2]
00:06:24.147  /usr/lib64/libclang-cpp.so.11(_ZN5clang8ParseASTERNS_4SemaEbb+0x2a5)[0x7f66ff3a1285]
00:06:24.147  /usr/lib64/libclang-cpp.so.11(_ZN5clang14FrontendAction7ExecuteEv+0xb9)[0x7f670078bb99]
00:06:24.147  /usr/lib64/libclang-cpp.so.11(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x1e3)[0x7f670075e793]
00:06:24.147  fatal error: error in backend: IO failure on output stream: No space left on device
00:06:24.147  /usr/lib64/libclang-cpp.so.11(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x6d0)[0x7f67007ee400]
00:06:24.147  /usr/bin/clang++(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x910)[0x555d416faae0]
00:06:24.147  /usr/bin/clang++(+0x198cd)[0x555d416fd8cd]
00:06:24.148  /usr/lib64/libclang-cpp.so.11(+0x1a26d29)[0x7f6700497d29]
00:06:24.148  /usr/lib64/libLLVM-11.so(_ZN4llvm20CrashRecoveryContext9RunSafelyENS_12function_refIFvvEEE+0x23)[0x7f66fa138b53]
00:06:24.148  /usr/lib64/libclang-cpp.so.11(_ZNK5clang6driver10CC1Command7ExecuteEN4llvm8ArrayRefINS2_8OptionalINS2_9StringRefEEEEEPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPb+0x157)[0x7f67004ab827]
00:06:24.148  /usr/lib64/libclang-cpp.so.11(_ZNK5clang6driver11Compilation14ExecuteCommandERKNS0_7CommandERPS3_+0x9b)[0x7f67004b1e0b]
00:06:24.149  /usr/lib64/libclang-cpp.so.11(_ZNK5clang6driver11Compilation11ExecuteJobsERKNS0_7JobListERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x7b)[0x7f67004b21db]
00:06:24.149  /usr/lib64/libclang-cpp.so.11(_ZN5clang6driver6Driver18ExecuteCompilationERNS0_11CompilationERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0xeb)[0x7f67004b356b]
00:06:24.149  /usr/bin/clang++(main+0x2277)[0x555d416f7007]
00:06:24.149  /usr/lib64/libc.so.6(__libc_start_main+0xf2)[0x7f66f93d31e2]
00:06:24.149  /usr/bin/clang++(_start+0x2e)[0x555d416f817e]
00:06:24.152  make[2]: *** [/home/vagrant/spdk_repo/spdk/mk/spdk.common.mk:462: nvmf_fc_spec.o] Error 1
00:06:24.152  make[2]: *** Waiting for unfinished jobs....
00:06:24.160  PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.

and the Backtrace:

00:06:29.393  ========== Backtrace start: ==========
00:06:29.393  
00:06:29.402  in /home/vagrant/spdk_repo/spdk/autobuild.sh:289 -> unittest_build([])
00:06:29.402       ...
00:06:29.402     284 		run_test "autobuild_build_doc" build_doc
00:06:29.402     285 	}
00:06:29.402     286 	
00:06:29.402     287 	function unittest_build() {
00:06:29.402     288 		"$rootdir/configure" $config_params --without-shared
00:06:29.402  => 289 		$MAKE $MAKEFLAGS
00:06:29.402     290 	}
00:06:29.402     291 	
00:06:29.402     292 	if [ $SPDK_RUN_VALGRIND -eq 1 ]; then
00:06:29.402     293 		run_test "valgrind" echo "using valgrind"
00:06:29.402     294 	fi
00:06:29.403       ...
00:06:29.411  in /home/vagrant/spdk_repo/spdk/test/common/autotest_common.sh:957 -> run_test(["unittest_build"],["unittest_build"])
00:06:29.411       ...
00:06:29.411     952 		timing_enter $test_name
00:06:29.411     953 		echo "************************************"
00:06:29.411     954 		echo "START TEST $test_name"
00:06:29.411     955 		echo "************************************"
00:06:29.411     956 		xtrace_restore
00:06:29.411  => 957 		time "$@"
00:06:29.411     958 		xtrace_disable
00:06:29.411     959 		echo "************************************"
00:06:29.411     960 		echo "END TEST $test_name"
00:06:29.411     961 		echo "************************************"
00:06:29.411     962 		timing_exit $test_name
00:06:29.412       ...
00:06:29.421  in /home/vagrant/spdk_repo/spdk/autobuild.sh:322 -> main(["/home/vagrant/autorun-spdk.conf"])
00:06:29.421       ...
00:06:29.421     317 	fi
00:06:29.421     318 	
00:06:29.421     319 	if [[ $SPDK_TEST_AUTOBUILD -eq 1 ]]; then
00:06:29.421     320 		run_test "autobuild" autobuild_test_suite $1
00:06:29.421     321 	elif [[ $SPDK_TEST_UNITTEST -eq 1 ]]; then
00:06:29.421  => 322 		run_test "unittest_build" unittest_build
00:06:29.421     323 	elif [[ $SPDK_TEST_SCANBUILD -eq 1 ]]; then
00:06:29.421     324 		run_test "scanbuild_make" scanbuild_make
00:06:29.421     325 	else
00:06:29.421     326 		# if we aren't testing the unittests, build with shared objects.
00:06:29.421     327 		./configure $config_params --with-shared
00:06:29.422       ...
00:06:29.422  
00:06:29.422  ========== Backtrace end ==========

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @KKaras169169. log: https://ci.spdk.io/public_build/autotest-per-patch_56527.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @KKaras169169. log: https://ci.spdk.io/public_build/autotest-per-patch_56716.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @KKaras169169. log: https://ci.spdk.io/public_build/autotest-per-patch_56730.html

**Comment by mikeBashStuff** [comment]
Currently this seems to be affecting only this series: https://review.spdk.io/gerrit/c/spdk/spdk/+/9241. I would consider testing it locally  to determine how big the clang build gets - the CI's VMs are very compact in nature and they don't have much of a storage space. It can be extended but first we need to know if it's really needed (and how much space is actually required to accommodate for this change).

**Comment by tomzawadzki** [comment]
@KKaras169169 will check the clang build size to determine if the VMs storage should be increased, or there is another issue present.

**Comment by KKaras169169** [comment]
Tested this locally (used "watch du -sh spdk/" command) and here are the results:

before build - 307 MB

with the series after build - 7,4 GB
without my series after build - 744 MB

Something is bloating the build to this enormous size. I am going to keep testing it to find the source.

**Comment by KKaras169169** [comment]
I have found a few things while comparing the builds sizes:

with configuration:
./configure --enable-debug --enable-werror --with-rdma --with-idxd --with-iscsi-initiator --with-nvme-cuse --with-pmdk --with-reduce --with-rbd --with-crypto --with-ocf --enable-ubsan --enable-asan --with-fuse --with-raid5 --without-shared
(I had to disable coverage, because I got errors with llvm library)

branch spdk-0.54:

- the archive file /spdk/intel-ipsec-mb/libIPSec_MB.a - 27M
- total size of our build/ directory - 1.6G
- total size of our test/ directory - 1.7G
overall size of the spdk - **3.8G**

branch spdk-1.0:

- the archive file /spdk/intel-ipsec-mb/lib/libIPSec_MB.a - 49M
- total size of our build/bin directory - 2.2G
- total size of our test/ directory - 2.4G
overall size of the spdk - **5.2G**

Even without all the configuration options enabled, we can see that with debug options active, our binaries increase in size significantly.

**Comment by jimharris** [comment]
maybe we should consider using shared library builds on systems with crypto enabled?

**Comment by tomzawadzki** [comment]
[Bug scrub] Suggestion from Jim is the right one, currently 3 jobs use --without-shared and there is no need to do so. @KKaras169169 will take a look at using shared build on clang job.

**Comment by KKaras169169** [comment]
This [patch](https://review.spdk.io/gerrit/c/spdk/spdk/+/9479/3) seems to fix the problem.

**Comment by jimharris** [comment]
I'm not sure, but would it be better to use a build parameter to control shared v. static instead of tying it directly to clang?

So with gcc, you see the static builds come out much smaller?

**Comment by KKaras169169** [comment]
1) A parameter would be much better for when we run into similar issues with expanding submodules.
2) Today I tested this with branch spdk-1.0 of intel-ipsec-mb checked out and --without-shared option enabled during the ./configure stage, and gcc build comes out about 0,6 GB smaller.

**Comment by tomzawadzki** [comment]
[Bug scrub] Comment from @jimharris is correct that differentiating only by the compiler within the test scripts might not provide the best control. It seems that we have an option to increase the infrastructure and VM disk size to accommodate the build sizes. Let's go this route if possible.

**Comment by tomzawadzki** [comment]
The size increase of the VM disk size is scheduled for deployment, @mikeBashStuff please close the issue once that is done.

**Comment by KKaras169169** [comment]
We have got a verification from CI on the [intel-ipsec-mb update](https://review.spdk.io/gerrit/c/spdk/spdk/+/7976)after the size has been increased. Closing the issue.
